qiskit_optimization/translators/docplex_mp.py

```diff
@@ -0,0 +1,286 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Translator between a docplex.mp model and a quadratic program"""
+
+from typing import TYPE_CHECKING, cast
+
+from docplex.mp.constr import LinearConstraint as DocplexLinearConstraint
+from docplex.mp.constr import NotEqualConstraint
+from docplex.mp.constr import QuadraticConstraint as DocplexQuadraticConstraint
+from docplex.mp.dvar import Var
+from docplex.mp.model import Model
+from docplex.mp.quad import QuadExpr
+from docplex.mp.vartype import BinaryVarType, ContinuousVarType, IntegerVarType
+
+from qiskit_optimization.exceptions import QiskitOptimizationError
+from qiskit_optimization.problems.constraint import Constraint
+from qiskit_optimization.problems.quadratic_objective import QuadraticObjective
+from qiskit_optimization.problems.variable import Variable
+
+if TYPE_CHECKING:
+    # pylint: disable=cyclic-import
+    from qiskit_optimization.problems.quadratic_program import QuadraticProgram
+
+
+def to_docplex_mp(quadratic_program: "QuadraticProgram") -> Model:
+    """Returns a docplex.mp model corresponding to a quadratic program.
+
+    Args:
+        quadratic_program: The quadratic program to be translated.
+
+    Returns:
+        The docplex.mp model corresponding to a quadratic program.
+
+    Raises:
+        QiskitOptimizationError: if non-supported elements (should never happen).
+    """
+    # initialize model
+    mdl = Model(quadratic_program.name)
+
+    # add variables
+    var = {}
+    for idx, x in enumerate(quadratic_program.variables):
+        if x.vartype == Variable.Type.CONTINUOUS:
+            var[idx] = mdl.continuous_var(lb=x.lowerbound, ub=x.upperbound, name=x.name)
+        elif x.vartype == Variable.Type.BINARY:
+            var[idx] = mdl.binary_var(name=x.name)
+        elif x.vartype == Variable.Type.INTEGER:
+            var[idx] = mdl.integer_var(lb=x.lowerbound, ub=x.upperbound, name=x.name)
+        else:
+            # should never happen
+            raise QiskitOptimizationError("Unsupported variable type: {}".format(x.vartype))
```

Comment:
I wonder whether it would be better to just throw a RuntimeError given this "should never happen". If it does its some sort of coding error or whatever - in some internal error that the user most likely cannot do anything about to overcome this.  So perhaps the message could be "Internal error: unsupported variable type:. {}"

Same for contrainst below.

Reply:
You are right. This is internal error that should not happen. So, It makes sense to change it `RuntimeError("Internal error: ...")` for `to_docplex_mp`.